draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+As part of a [presentation](https://github.com/troessner/talks/blob/master/exploiting_rubys_ast_a_love_story_in_three_chapters/presentation.html) I'm hoping to give at the end of this year about abstract syntax trees and how you can leverage them I started to look into how the [Mutant](https://github.com/mbj/mutant) gem works. `Mutant` is the by far most advanced gem for doing mutation testing in Ruby. I was (and still am) amazed how incredibly well built it is on all levels, ranging from the "big picture" architecture down to the "nuts and bolts" source code.
+
+When it comes to mutation testing I learned one thing. It's complex. Really complex. And as a consequence, `Mutant` is complex. 
+
+I did learn a lot from reading its source and I'd like to share this with you.
+
+**Beware though, this is going to be a long ride.**
+
+ This article will probably take you at least half an hour to an hour. But at the end, it'll hopefully be worth your while.
+
+### Mutation testing in a nutshell
+
+Let's quickly talk about what mutation testing is first. Mutation testing takes code like this:
+
+```Ruby
+class Greeter
+  def initialize(phrase, enabled = true)
+    @phrase = phrase
+    @enabled = enabled
+  end
+
+  def say_hello(name)
+    "#{@phrase} #{name}" if @enabled
+  end
+end
+```
+
+and turns it into something like this:
+
+```Ruby
+def say_hello(name)
+  if true
+    "#{@phrase}#{" "}#{name}"
+  end
+end
+```
+
+or 
+
+```Ruby
+def say_hello(name)
+  if false
+    "#{@phrase}#{" "}#{name}"
+  end
+end
+```
+
+`Mutant` then runs your tests against each mutation. The tests might look like this:
```

Comment:
>>  I think you cannot go into the details of the test selection and the various trade offs here.

I actually have not looked into this at all but for sure I'd be interested in covering this maybe in a follow up blog post.
Changed it to "tests" now.